webots_ros2_driver/src/plugins/static/Ros2Compass.cpp

```diff
@@ -0,0 +1,110 @@
+      RCLCPP_DEBUG(rclcpp::get_logger(mDeviceName), (mDeviceName + " activated!").c_str());
+      mIsEnabled = true;
+    }
+  }
+  void Ros2Compass::step() {
+    if (!preStep())
+      return;
+
+    if (mIsEnabled)
+      publishData();
+
+    if (mAlwaysOn)
+      return;
+
+    mustPublish();
+  }
+  bool Ros2Compass::mustPublish() {
+    // Enable/Disable sensor
+    const bool shouldBeEnabled = mDataPublisher->get_subscription_count() > 0;
+    if (shouldBeEnabled != mIsEnabled) {
+      if (shouldBeEnabled)
+        wb_compass_enable(mCompass, mPublishTimestepSyncedMs);
+      else
+        wb_compass_disable(mCompass);
+      mIsEnabled = shouldBeEnabled;
+    }
+    return mIsEnabled;
+  }
+  void Ros2Compass::publishData() {
+    const double *compass_values = wb_compass_get_values(mCompass);
+    mDataMessage.header.stamp = mNode->get_clock()->now();
+    mDataMessage.magnetic_field.x = compass_values[0];
+    mDataMessage.magnetic_field.y = compass_values[1];
+    mDataMessage.magnetic_field.z = compass_values[2];
+    mDataPublisher->publish(mDataMessage);
+  }
+  void Ros2Compass::enable_callback(const std::shared_ptr<webots_ros2_msgs::srv::SetInt::Request> request,
+                                    std::shared_ptr<webots_ros2_msgs::srv::SetInt::Response> response) {
+    if (request->value) {
+      wb_compass_enable(mCompass, mPublishTimestepSyncedMs);
+      mAlwaysOn = true;
+      mIsEnabled = true;
+    }
+    response->success = true;
+  }
+  void Ros2Compass::disable_callback(const std::shared_ptr<webots_ros2_msgs::srv::SetInt::Request> request,
+                                     std::shared_ptr<webots_ros2_msgs::srv::SetInt::Response> response) {
+    if (request->value) {
+      wb_compass_disable(mCompass);
+      mAlwaysOn = false;
+      mIsEnabled = false;
+    }
+    response->success = true;
+  }
+  void Ros2Compass::get_sampling_period_callback(const std::shared_ptr<webots_ros2_msgs::srv::GetInt::Request> request,
+                                                 std::shared_ptr<webots_ros2_msgs::srv::GetInt::Response> response) {
+    response->value = request->ask ? wb_compass_get_sampling_period(mCompass) : 0;
+  }
+  void Ros2Compass::get_lookup_table_callback(const std::shared_ptr<webots_ros2_msgs::srv::GetFloatArray::Request> request,
+                                              std::shared_ptr<webots_ros2_msgs::srv::GetFloatArray::Response> response) {
+    int table_size = wb_compass_get_lookup_table_size(mCompass);
+    std::vector<double> values(table_size * 3);
+    if (request->ask) {
+      const double *lookup_table = wb_compass_get_lookup_table(mCompass);
+      for (int i = 0; i < table_size; i++) {
+        int offset = i * 3;
+        values.at(offset) = lookup_table[offset];
+        values.at(offset + 1) = lookup_table[offset + 1];
+        values.at(offset + 2) = lookup_table[offset + 2];
+      }
+    }
+    response->values = values;
```

Comment:
```suggestion
    const double *lookup_table = wb_compass_get_lookup_table(mCompass);
    for (int i = 0; i < table_size; i++) {
      int offset = i * 3;
      values.at(offset) = lookup_table[offset];
      values.at(offset + 1) = lookup_table[offset + 1];
      values.at(offset + 2) = lookup_table[offset + 2];
    }
    response->value = values;
```

We can get rid of the ask input. This is not mandatory.